osu.Game/Online/Rooms/OnlinePlayBeatmapAvailabilityTracker.cs

```diff
@@ -40,71 +50,67 @@ protected override void LoadComplete()
                 if (item.NewValue == null)
                     return;
 
-                Model.Value = item.NewValue.Beatmap.Value.BeatmapSet;
-            }, true);
+                downloadTracker?.RemoveAndDisposeImmediately();
 
-            Progress.BindValueChanged(_ =>
-            {
-                if (State.Value != DownloadState.Downloading)
-                    return;
-
-                // incoming progress changes are going to be at a very high rate.
-                // we don't want to flood the network with this, so rate limit how often we send progress updates.
-                if (progressUpdate?.Completed != false)
-                    progressUpdate = Scheduler.AddDelayed(updateAvailability, progressUpdate == null ? 0 : 500);
-            });
+                downloadTracker = new BeatmapDownloadTracker(item.NewValue.Beatmap.Value.BeatmapSet);
+                downloadTracker.State.BindValueChanged(_ => updateAvailability());
+                downloadTracker.Progress.BindValueChanged(_ =>
+                {
+                    if (downloadTracker.State.Value != DownloadState.Downloading)
+                        return;
 
-            State.BindValueChanged(_ => updateAvailability(), true);
-        }
+                    // incoming progress changes are going to be at a very high rate.
+                    // we don't want to flood the network with this, so rate limit how often we send progress updates.
+                    if (progressUpdate?.Completed != false)
+                        progressUpdate = Scheduler.AddDelayed(updateAvailability, progressUpdate == null ? 0 : 500);
+                });
 
-        protected override bool VerifyDatabasedModel(BeatmapSetInfo databasedSet)
-        {
-            int beatmapId = SelectedItem.Value?.Beatmap.Value.OnlineID ?? -1;
-            string checksum = SelectedItem.Value?.Beatmap.Value.MD5Hash;
-
-            var matchingBeatmap = databasedSet.Beatmaps.FirstOrDefault(b => b.OnlineBeatmapID == beatmapId && b.MD5Hash == checksum);
-
-            if (matchingBeatmap == null)
-            {
-                Logger.Log("The imported beatmap set does not match the online version.", LoggingTarget.Runtime, LogLevel.Important);
-                return false;
-            }
-
-            return true;
-        }
-
-        protected override bool IsModelAvailableLocally()
-        {
-            int onlineId = SelectedItem.Value.Beatmap.Value.OnlineID;
-            string checksum = SelectedItem.Value.Beatmap.Value.MD5Hash;
-
-            var beatmap = Manager.QueryBeatmap(b => b.OnlineBeatmapID == onlineId && b.MD5Hash == checksum);
-            return beatmap?.BeatmapSet.DeletePending == false;
+                AddInternal(downloadTracker);
```

Comment:
Is this correct...? The `AddInternal()` being inside the progress bindable's callback? Care to explain how this works?

Reply:
I think the diff might be confusing you here, github being a bit weird it'd seem.

Reply:
Oh, I see. There's two nested `BindValueChange()`s...